hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyDeletingService.java

```diff
@@ -234,27 +224,126 @@ private Pair<Integer, Boolean> submitPurgeKeysRequest(List<DeleteBlockGroupResul
-        .setPurgeKeysRequest(purgeKeysRequest)
-        .setClientId(getClientId().toString())
-        .build();
-
-    // Submit PurgeKeys request to OM. Acquire bootstrap lock when processing deletes for snapshots.
-    try (BootstrapStateHandler.Lock lock = snapTableKey != null ? getBootstrapStateLock().lock() : null) {
-      OzoneManagerProtocolProtos.OMResponse omResponse = submitRequest(omRequest);
-      if (omResponse != null) {
-        purgeSuccess = purgeSuccess && omResponse.getSuccess();
+      // Now actually add batchPurgeKeys
+      if (!batchPurgeKeys.isEmpty()) {
+        OzoneManagerProtocolProtos.DeletedKeys deletedKeys = OzoneManagerProtocolProtos.DeletedKeys.newBuilder()
+            .setVolumeName("").setBucketName("")
+            .addAllKeys(batchPurgeKeys).build();
+        requestBuilder.addDeletedKeys(deletedKeys);
+        currSize = requestBuilder.build().getSerializedSize();
+      }
+
+      remainingRatisLimit -= currSize;
+
+      // 3.2 Add keysToUpdate
+      while (!batchCapacityReached && updateIndex < keysToUpdateList.size()) {
+        OzoneManagerProtocolProtos.SnapshotMoveKeyInfos nextUpdate = keysToUpdateList.get(updateIndex);
+
+        PurgeKeysRequest.Builder tempBuilder = requestBuilder.clone();
+        tempBuilder.addKeysToUpdate(nextUpdate);
+        int estimatedSize = tempBuilder.build().getSerializedSize();
+
+        if (currSize + estimatedSize > remainingRatisLimit) {
+          batchCapacityReached = true;
+          break;
+        }
+
+        requestBuilder.addKeysToUpdate(nextUpdate);
+        currSize += estimatedSize;
+        updateIndex++;
+      }
+
+      remainingRatisLimit -= currSize;
+
+      // 3.3 Add renamed keys
+      while (!batchCapacityReached && renameEntriesToBeDeleted != null &&
+          renameIndex < renameEntriesToBeDeleted.size()) {
+        String nextRename = renameEntriesToBeDeleted.get(renameIndex);
+
+        PurgeKeysRequest.Builder tempBuilder = requestBuilder.clone();
+        tempBuilder.addRenamedKeys(nextRename);
+        int estimatedSize = tempBuilder.build().getSerializedSize();
+
+        if (currSize + estimatedSize > remainingRatisLimit) {
+          break;
+        }
+
+        requestBuilder.addRenamedKeys(nextRename);
+        currSize += estimatedSize;
+        renameIndex++;
+      }
+
+      // Finalize and send this batch
+      OzoneManagerProtocolProtos.OMRequest omRequest = OzoneManagerProtocolProtos.OMRequest.newBuilder()
+          .setCmdType(OzoneManagerProtocolProtos.Type.PurgeKeys)
+          .setPurgeKeysRequest(requestBuilder.build())
+          .setClientId(getClientId().toString())
+          .build();
+
+      try (BootstrapStateHandler.Lock lock =
+          snapTableKey != null ? getBootstrapStateLock().lock() : null) {
+        OzoneManagerProtocolProtos.OMResponse omResponse = submitRequest(omRequest);
+        if (omResponse != null) {
+          purgeSuccess = purgeSuccess && omResponse.getSuccess();
```

Comment:
since batch is divided, do previous failure can have some impact here? if batch is not stopped and continue.

Additionally, do any relation exist between purgeKeys, renameEntriesToBeDeleted and keysToUpdateList
Earlier no batch, so its atomic operation. Now like purgeKeys success and system crash. And startup, will it able to determine other 2 list ?